### PR TITLE
Prologue Fusion

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134532
	

This PR extends our ability to fuse pointwise nodes onto triton templates with the ability to fuse pointwise nodes into triton templates - prologue fusion.

Similar to the store_output api:
`{{store_output(("idx_m", "idx_n"), "acc", "mask")}}`

And the modification api:

```
{{ modification(
    subgraph_number=0,
    output_name="post_mod_scores",
    score="qk",
    out="qk"
) | indent_except_first(1) }}
```

We have:

```{{load_input("B", "b", ("idx_m", "idx_n"), mask=None if EVEN_K else "b_mask", indent_width=8)}}```

Because we are now loading the input with explicit indices and mask, I needed to rewrite the mm kernel to no longer update the [pointers by BLOCK_K](https://github.com/pytorch/pytorch/blob/bb03ef7acadf7bbc3287b0ada58e9476eda6e0fe/torch/_inductor/kernel/mm.py#L110-L111) on every iteration and instead on each iteration compute indices from the the k_idx of each loop. This did not have any perf difference.

There are a couple main use cases for prologue fusion:

- Fusing dequants into a matmul. particularly for more bandwidth bound scenarios.
- Fusing gather into a matmul. This is useful particularly in MOE. See https://github.com/pytorch/pytorch/issues/134535 for more details.

Prologue fusion is generally much less profitable than epilogue fusion, because it must be applied to an element of an input on each loop of the matmul, compared to only once in the epilogue (gather into matmul is a potential exception). Accordingly, we are much less aggressive in attempting to fuse prologue fusion. We only attempt fusion if it does not increase the number of memory bytes read instead the triton template, multipled by a small factor to allow gathers. This restricts reliably unprofitable fusions like fp32->fp16 inside kernel. In future pr we could potentially have api of being more aggressive if we know we are in a bandwidth bound regime. See: https://github.com/pytorch/pytorch/pull/134532/files#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7R3060-R3066

Other notes:

By default we will upcast to fp32 inside every kernel. This matches eager numerics. This is fine enough for epilogue because it is only done once (although it is probably unnecessary for say a relu) but tanks perf for prologue. I am currently using the `codegen_upcast_to_fp32` option to avoid it, but that will not work for libdevice calls that require fp32. We will need https://github.com/pytorch/pytorch/pull/136778/ and dtype-aware codegen to upcast fp16 ops into libdevice calls.

With prologue fusion, we now have essentially separate kernels for each input, and for the output. I had to increase the number of fields that are swapped out in `set_subgraph_body` by a large number :/ I also update the fusion logic because the inputs will have a different group than the outputs. Maybe as part of enabling multiple outputs, this could get cleaned up a bit so.. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/_inductor/codegen/cuda_combined_scheduling.py [hunk 1]
```diff
@@ -60,20 +60,23 @@ def codegen_template(
         self,
         template_node: BaseSchedulerNode,
         epilogue_nodes: Sequence[BaseSchedulerNode],
+        prologue_nodes: Sequence[BaseSchedulerNode],
     ):
         if self._cuda_cpp_scheduling.is_cuda_cpp_template(template_node):
             assert epilogue_nodes is None or len(epilogue_nodes) == 0
+            assert prologue_nodes is None or len(epilogue_nodes) == 0
```
Comment (jansel) [inline torch/_inductor/codegen/cuda_combined_scheduling.py:67]:
```suggestion
            assert prologue_nodes is None or len(prologue_nodes) == 0
```
or even
```suggestion
            assert not prologue_nodes
```

File: torch/_inductor/codegen/cuda_combined_scheduling.py [hunk 2]
```diff
@@ -60,20 +60,23 @@ def codegen_template(
...
             return self._cuda_cpp_scheduling.codegen_template(
                 template_node, epilogue_nodes
             )
         elif self._rocm_cpp_scheduling.is_rocm_cpp_template(template_node):
             assert epilogue_nodes is None or len(epilogue_nodes) == 0
+            assert prologue_nodes is None or len(epilogue_nodes) == 0
```
Comment (jansel) [inline torch/_inductor/codegen/cuda_combined_scheduling.py:73]: same

File: torch/_inductor/codegen/simd.py [hunk 3]
```diff
@@ -1498,28 +1531,69 @@ def codegen_template(
...
         assert rnumel == 1
         kernel, render = template_node.node.make_kernel_render(template_node.node)
+
+        buf_name_to_prologue_group = {}
+        template_reads = template_node.used_buffer_names()
+        prologue_group = []
+        for prologue in prologue_nodes:
+            names = prologue.get_buffer_names()
+            prologue_group.append(prologue)
+            if prologue.get_buffer_names() & template_reads:
```
Comment (jansel) [inline torch/_inductor/codegen/simd.py:1541]: If this is False what happens?  Should that be an error?

Comment (jansel) [inline torch/_inductor/codegen/simd.py:1541]:
```suggestion
            if names & template_reads:
```

File: torch/_inductor/kernel/mm.py [hunk 4]
```diff
@@ -85,30 +85,35 @@
...
+        {% if not EVEN_K %}
+        a_mask = offs_k[None, :] < (K - k_idx * BLOCK_K)
+        b_mask = offs_k[:, None] < (K - k_idx * BLOCK_K)
+        {% endif %}
+        a_k_idx_vals = offs_k[None, :] + (k_idx * BLOCK_K)
+        b_k_idx_vals = offs_k[:, None] + (k_idx * BLOCK_K)
+
+        idx_m = offs_a_m[:, None]
+        idx_n = a_k_idx_vals
+        {{load_input("A", "a", ("idx_m", "idx_n"), mask=None if EVEN_K else "a_mask", indent_width=8)}}
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:108]: Do we have some sort of assert that the template is using this new API?

I want to make sure that for other templates (or user-define ones) we don't do prologue fusion and the template never calls this.

File: torch/_inductor/kernel/mm_common.py [hunk 5]
```diff
@@ -121,24 +121,24 @@ def filtered_configs(
...
+        # {"config": (64, 128, 32, 3, 4), "cond": True},
+        # {"config": (64, 128, 32, 4, 8), "cond": True},
+        # {"config": (64, 128, 64, 3, 4), "cond": True},
+        # {"config": (64, 128, 128, 4, 4), "cond": True},
+        # {"config": (128, 64, 32, 3, 4), "cond": True},
+        # {"config": (128, 64, 32, 4, 8), "cond": True},
+        # {"config": (128, 128, 32, 2, 8), "cond": True},
+        # {"config": (128, 128, 32, 3, 4), "cond": True},
+        # {"config": (128, 128, 64, 3, 4), "cond": True},
+        # {"config": (128, 128, 64, 5, 8), "cond": True},
```
Comment (jansel) [inline torch/_inductor/kernel/mm_common.py:141]: ?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/simd.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/codegen/simd.py]: If the prologue node does not have any overlap, that means that there are subsequent prologue nodes to come that must. I added an assert below that all groups will end up with overlapping names of template.

It's probably not particularly likely and only happens when you force buffer realization for pointwise nodes... we can clean up a bit when we support max-autotune multiple outputs

File: torch/_inductor/kernel/mm.py (hunk 4 above)
Comment (eellison) [reply torch/_inductor/kernel/mm.py:109]: When we generate the kernel, we record which names have been loaded with this api and only allow prologue fusion on those inputs.

**Review by jansel — APPROVED** [review]

**Comment by nullplay** [comment]
@eellison

Hi, could you provide an update on the status of this PR? I believe it would be really useful for my use case. 
Also, is there support for epilogue fusion of pointwise operators with the Triton matmul template as well?

**Comment by clee2000** [comment]
@pytorchbot revert -m "A couple of PRs in this stack are breaking internally on different tests" -c ghfirst

https://github.com/pytorch/pytorch/pull/134532 D67035420 - hardware not in oss ci?
https://github.com/pytorch/pytorch/pull/142350 D67035434 - hardware not in oss ci?
https://github.com/pytorch/pytorch/pull/142435 D67056571 - fails the tests added on the first PR in this stack

**Comment by clee2000** [comment]
@pytorchbot revert -m "A couple of PRs in this stack are breaking internally on different tests" -c ghfirst

https://github.com/pytorch/pytorch/pull/134532 D67035420 - hardware not in oss ci?
https://github.com/pytorch/pytorch/pull/142350 D67035434 - hardware not in oss ci?
https://github.com/pytorch/pytorch/pull/142435 D67056571 - fails the tests added on the first PR in this stack

**Pull request closed without merging**